### ISSUE TITLE
Client handlers break when loading and immediately attaching to buffer

### ISSUE THREAD

**Issue by jose-elias-alvarez** [neovim/neovim#16562]
### Neovim version (nvim -v)

NVIM v0.7.0-dev+683-gafaad8b54

### Language server name/version

pylsp v1.3.2 (but not relevant to issue)

### Operating system/version

macOS 12.0.1

### :checkhealth

```markdown
lspconfig: require("lspconfig.health").check()
========================================================================
  - INFO: sumneko_lua: configuration checked.
  - INFO: denols: configuration checked.
  - INFO: tsserver: configuration checked.
  - INFO: jsonls: configuration checked.
  - INFO: bashls: configuration checked.
  - INFO: pylsp: configuration checked.
```

### Steps to reproduce using "nvim -u minimal_init.lua"

All that's needed is the following config:

```lua
-- minimal_init.lua
vim.lsp.set_log_level("trace")

-- cmd should correspond to an installed server but is otherwise irrelevant
local client_id = vim.lsp.start_client({ cmd = { "pylsp" } }) 
vim.lsp.buf_attach_client(vim.api.nvim_get_current_buf(), client_id)
```

Then, open Neovim with `nvim --clean -u minimal_init.lua test.py` (the filename is irrelevant).

### Expected behavior

The client sends notifications on buffer events (e.g. `textDocument/didChange` on change). 

### Actual behavior

The client sends a `textDocument/didOpen` notification but does not send any further notifications on events. 

I think the issue is that the buffer is not yet loaded when `vim.lsp.buf_attach_client` is called, so [`vim.api.nvim_buf_attach`](https://github.com/neovim/neovim/blob/9b65a7e06438c3b407ad7136a30c0a5281f2174f/runtime/lua/vim/lsp.lua#L1173) fails and returns `false`. However, that return value isn't checked, so the attach process fails silently (which can affect subsequent servers, too, since the client is under the impression that the buffer is already attached). 

There's a couple of simple solutions on the user / plugin side, like using an autocommand that fires after the buffer is loaded or deferring `vim.lsp.buf_attach_client`, but I wonder if there's a way to at least indicate that the process failed in this case (and ideally avoid breaking other servers). 

### Log file

https://gist.github.com/jose-elias-alvarez/0c2e46f2db52a3dee2afca3ccceca7d8

**Comment by mfussenegger** [comment]
Can you elaborate in which case you would call `vim.lsp.buf_attach_client` before the buffer is ready?
Currently the `start_client` / `buf_attach_client` functions are mostly used by lspconfig and a couple of language specific extensions where they're mostly triggered as part of a FileType event where this isn't an issue.

(Not saying we shouldn't fix this, but this could influence how we fix it)

**Comment by jose-elias-alvarez** [comment]
The issue is a bit hard to understand, but this came up in jose-elias-alvarez/null-ls.nvim#401. To try to sum it up, this is what happened:

1. The user sets up null-ls, which normally checks whether it should attach in a `FileType` autocommand.
2. The user opens Neovim from the command line with a file as an argument, as in the example here.
3. The user's config immediately registers a null-ls diagnostics source, which triggers an independent check to see if any open buffer(s) should attach.
4. The check succeeds, and the error described in this issue is triggered.

Basically, I think this could happen in cases where the user or a plugin has code that calls `buf_attach_client` using conditions that could be met as soon as a file is opened.

Doing this in the first place is probably not a good idea (and we're working on refactoring null-ls to avoid the issue), but it would be really helpful to at least have an indicator that attachment failed.

**Comment by mjlbach** [comment]
Please test https://github.com/neovim/neovim/pull/16723
